src/EditorFeatures/Core/Implementation/FindReferences/AbstractFindReferencesService.cs

```diff
@@ -51,36 +59,127 @@ protected AbstractFindReferencesService(IEnumerable<IReferencedSymbolsPresenter>
             return null;
         }
 
-        public async Task<IEnumerable<INavigableItem>> FindReferencesAsync(Document document, int position, IWaitContext waitContext)
+        /// <summary>
+        /// Finds references using the externally defined <see cref="IFindReferencesResultProvider"/>s.
+        /// </summary>
+        private async Task AddExternalReferencesAsync(Document document, int position, ArrayBuilder<INavigableItem> builder, CancellationToken cancellationToken)
+        {
+            if (_externalReferencesProviders.Any())
+            {
+                foreach (var provider in _externalReferencesProviders)
+                {
+                    var references = await provider.FindReferencesAsync(document, position, cancellationToken).ConfigureAwait(false);
+                    if (references != null)
+                    {
+                        builder.AddRange(references.WhereNotNull());
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        /// Finds references using <see cref="SymbolFinder.FindReferencesAsync(ISymbol, Solution, CancellationToken)"/>
+        /// </summary>
+        private async Task AddSymbolReferencesAsync(Document document, int position, ArrayBuilder<INavigableItem> builder, IWaitContext waitContext)
         {
-            var cancellationToken = waitContext.CancellationToken;
             var result = await this.FindReferencedSymbolsAsync(document, position, waitContext).ConfigureAwait(false);
-            if (result == null)
+            if (result != null)
             {
-                return SpecializedCollections.EmptyEnumerable<INavigableItem>();
+                var referencedSymbols = result.Item1;
+                var searchSolution = result.Item2;
```

Comment:
I have actually just refactored the previous code and moved it within this method - I see that [FindReferencedSymbolsAsync](http://source.roslyn.io/#Microsoft.CodeAnalysis.EditorFeatures/Implementation/FindReferences/AbstractFindReferencesService.cs,35) might end up mapping the symbol search to a different solution (probably for MAS case).